--- redmine-0.9.3-fixes/app/helpers/issues_helper.rb~	2010-02-28 11:28:05.000000000 +0100
+++ redmine-0.9.3-fixes/app/helpers/issues_helper.rb	2010-03-12 19:52:55.262371393 +0100
@@ -144,7 +144,7 @@
   def issues_to_csv(issues, project = nil)
     ic = Iconv.new(l(:general_csv_encoding), 'UTF-8')    
     decimal_separator = l(:general_csv_decimal_separator)
-    export = FCSV.generate(:col_sep => l(:general_csv_separator)) do |csv|
+    export = FCSV.generate(:col_sep => l(:general_csv_separator), :encoding => Encoding::UTF_8) do |csv|
       # csv header fields
       headers = [ "#",
                   l(:field_status), 
