diff -ur redmine-1.0.4/app/models/auth_source_ldap.rb redmine-1.0.4-ldap/app/models/auth_source_ldap.rb
--- redmine-1.0.4/app/models/auth_source_ldap.rb	2010-03-07 23:29:20.233868713 +0100
+++ redmine-1.0.4-ldap/app/models/auth_source_ldap.rb	2010-03-07 23:28:36.000000000 +0100
@@ -130,7 +130,13 @@
   
   def self.get_attr(entry, attr_name)
     if !attr_name.blank?
-      entry[attr_name].is_a?(Array) ? entry[attr_name].first : entry[attr_name]
+      if entry[attr_name].is_a?(Array) 
+	 val = entry[attr_name].first
+      else
+	 val = entry[attr_name]
+      end
+      val.force_encoding(Encoding::UTF_8) if val.respond_to?(:force_encoding)
+      val
     end
   end
 end
