--- redmine-0.9.4/lib/redmine/scm/adapters/git_adapter.rb~	2010-12-02 12:25:36.877495204 +0100
+++ redmine-0.9.4/lib/redmine/scm/adapters/git_adapter.rb	2010-12-02 12:29:36.839495202 +0100
@@ -92,7 +92,7 @@
           shellout(cmd) do |io|
             begin
               id = lines[0].split[1]
-              author = lines[1].match('Author:\s+(.*)$')[1]
+              author = lines[1].force_encoding(Encoding::ASCII_8BIT).match('Author:\s+(.*)$')[1]
               time = Time.parse(lines[4].match('CommitDate:\s+(.*)$')[1])
 
               Revision.new({
@@ -130,6 +130,7 @@
             revno = 1
 
             io.each_line do |line|
+              line.force_encoding(Encoding::ASCII_8BIT) if line.respond_to?(:force_encoding)
               if line =~ /^commit ([0-9a-f]{40})$/
                 key = "commit"
                 value = $1
